From dc068d8365431e8d43029bf5184bcb9ca47a165f Mon Sep 17 00:00:00 2001
From: ztc1997 <dev@ztc1997.com>
Date: Sun, 1 Sep 2024 21:41:17 +0800
Subject: [PATCH] block: Do not allow boosters to adjusting scheduler

---
 block/elevator.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/block/elevator.c b/block/elevator.c
index e739f47e79cf..f4362fd7dc07 100644
--- a/block/elevator.c
+++ b/block/elevator.c
@@ -764,11 +764,16 @@ static int __elevator_change(struct request_queue *q, const char *name)
 	return elevator_switch(q, e);
 }
 
+bool task_is_booster(struct task_struct *tsk);
+
 ssize_t elv_iosched_store(struct request_queue *q, const char *name,
 			  size_t count)
 {
 	int ret;
 
+	if (task_is_booster(current))
+		return count;
+
 	if (!elv_support_iosched(q))
 		return count;
 
